design_docs/oscode/oscode_sundials.md

```diff
@@ -0,0 +1,375 @@
+
+------------------------------------
+
+The main driver of `OSCODE` will be the function `oscEvolve` with a similar signature to `arkEvolve`
+
+```C
+/*---------------------------------------------------------------
+  oscEvolve:
+
+  This routine is the main driver of OSCODE-based integrators.
+
+  It integrates over a time interval defined by the user, by
+  calling the time step module to do internal time steps.
+
+  The first time that oscEvolve is called for a successfully
+  initialized problem, it computes a tentative initial step size.
+
+  oscEvolve supports two modes as specified by itask: OSC_NORMAL and
+  OSC_ONE_STEP.  In the OSC_NORMAL mode, the solver steps until
+  it reaches or passes tout and then interpolates to obtain
+  y(tout).  In the OSC_ONE_STEP mode, it takes one internal step
+  and returns.  The behavior of both modes can be over-rided
+  through user-specification of osc_tstop (through the
+  *StepSetStopTime function), in which case if a solver step
+  would pass tstop, the step is shortened so that it stops at
+  exactly the specified stop time, and hence interpolation of
+  y(tout) is not required.
+  ---------------------------------------------------------------*/
+int oscEvolve(OSCodeMem osc_mem, realtype tout, N_Vector yout,
+              realtype *tret, int itask);
+```
+
+The `OSCodeMemRec` will be similar to `ARKodeMemRec` containing the information for the estimation using the rk and wkb solver
+
+(TODO:) The example program `example_oscode.cpp` shows an example of the user API for the burst equation.
+
+Users can set:
+
+1. `w` and `g` functions that take in a timestep `t` along with a `void*` argument for their own data
+2. `OSCODEOmegaFn` and `OSCODEGammaFn`, the functors $\omega(t)$ and $\gamma(t)$ which have pointer type
+    - `complextype (OSCODEOmegaFn*)(realtype)`
+3. Options for the rk solver
+    - order
+    - tolerances
+    - butcher tables
+    - Gauss - Lobatto weights
+    - Integral Function
+4. Options for the wkb solver
+    - Order
+    - order related weight matrices
+    - Derivative and series functions
+    - Function for Hermite and Lagrange
+    - Gauss-Lobatto weights
+5. Options for solver
+    - initial conditions for the ODE, $ x(t)  \frac{dx}{dt} $ evaluated at the start of the integration range
+    - start of integration range
+    - end of integration range
+    - do_times timepoints at which dense output is to be produced. Must be sorted
+    - Direction
+    - order of ARDC approximation to be used
+    - (local) relative tolerance
+    - (local) absolute tolerance
+    - initial stepsize to use
+
+The implimentation can utilize the ARKODE explicit runga kutta solver, but we will have to write an implementation of the ARDC solver. For a first implementation I think it would be fine to only have the 6th order ARDC solver already written in C++, then in another PR expose functions that allow the user to change the ARDC solvers order.
```

Comment:
ARDC is not fixed order, it's adaptive. Also it's not written in C++ -- not sure what happened here. It'll need to be translated from Python, but it's conceptually simple.

Reply:
Whoops yes, leftover from WKB